Sources/GraphQLWebSocket/Client.swift

```diff
@@ -0,0 +1,583 @@
+            retry = attempt
+        }
+        
+        self.health = .reconnecting(retry: retry + 1)
+        
+        self.config.logger.debug("Scheduling \(retry)-th connect retry!")
+        self.reconnectTimer = Timer.scheduledTimer(
+            withTimeInterval: TimeInterval(self.config.retryWait(retry)),
+            repeats: false,
+            block: { [weak self] _ in
+                // A reconnect should always happen if the client connects eagerly,
+                // even if we have no current listeners.
+                guard let self = self, !self.pipelines.isEmpty, self.config.behaviour != .eager else {
+                    return
+                }
+                
+                self.config.logger.debug("Reconnection in progress...")
+                self.connect()
+            })
+    }
+    
+    /// Sends a message using the websocket transport.
+    private func send(message: ClientMessage) {
+        guard let socket = self.socket else {
+            self.config.logger.debug("Socket not ready, queueing message \(message.description)...")
+            self.queue.append(message)
+            return
+        }
+        
+        let data = try! self.encoder.encode(message)
+        switch (self.health, message) {
+        case (.acknowledged, _), (.disposed, _), (_, .initialise):
+            // We can send any message when the connection has been ACK and meta messages when the server hasn't ACK the connection yet.
+            socket.write(data: data)
+            self.config.logger.debug("\(message.description) sent to the server!")
+            
+        default:
+            self.config.logger.debug("Transport not ready, queueing message \(message.description)...")
+            self.queue.append(message)
+        }
+    }
+    
+    /// Empties the message queue if possible.
+    private func flushQueue() {
+        // We take the count of messages in the queue before
+        // flushing it to make sure we only process each message once.
+        // Messages can reinsert themself in the queue if they are not yet ready.
+        var numberOfOutstandingMessages = self.queue.count
+        self.config.logger.debug("Flushing \(numberOfOutstandingMessages) messages from queue...")
+        
+        while numberOfOutstandingMessages > 0 {
+            self.send(message: self.queue.removeFirst())
+            numberOfOutstandingMessages -= 1
+        }
+        
+        self.config.logger.debug("Queue flushed!")
+    }
+    
+    /// Sends a ping request and starts the response timeout.
+    private func ping() {
+        self.send(message: ClientMessage.ping())
+        self.emitter.send(Event.ping(received: false, payload: nil))
+        self.config.logger.debug("Emitted a PING message!")
+        
+        // We schedule a response timeout that has to be cleared
+        // in a timely manner by receiveing a new message from the server.
+        self.connectionDroppedTimer = Timer.scheduledTimer(
```

Comment:
`ping()` is called every `config.keepAlive` seconds and each of these calls creates a new `connectionDroppedTimer` (destroying the current one) that waits for three times the keepAlive time. So only the last ping() call's `connectionDroppedTimer` can possibly ever return, is that intentional? 

Also I don't think it needs to repeat.

Reply:
`ping` is only called after the server and client haven't sent messages in a while (i.e. not every `config.keepAlive`). We want to fire (return) only when the timer times out and we haven't received any messages in the meantime.

Does that make sense?

Reply:
I understand, I think I must have misunderstood when I reviewed.